File: 90-subsets-ii/90-subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        res = []
        final = []
        def find(n):
            final.append(res[:])
            for i in range(n,len(nums)):
                if i > n and nums[i] == nums[i-1]:
                    continue
                res.append(nums[i])
                find(i+1)
                res.pop()
        find(0)
        return final

File: 90-subsets-ii/90-subsets-ii.java
class Solution {
    LinkedList<Integer> temp = new LinkedList<>();
    List<List<Integer>> res  = new LinkedList<>();
    Set<Integer> visited = new HashSet<>();

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        subset(nums, 0);
        return res;        
    }
    
    public void subset(int[] nums, int start){
        res.add((List) temp.clone());
        
        for(int i = start; i < nums.length; i++){
            int curNum = nums[i];
            if(i > 0 && nums[i-1] == nums[i] && !visited.contains(i-1))continue;
            visited.add(i);
            temp.addLast(curNum);
            subset(nums, i + 1);
            temp.pollLast();
            visited.remove(i);
        }
    }
    
}
